Source/OnlineSubsystemEpic/Private/OnlineFriendsInterfaceEpic.cpp

```diff
@@ -0,0 +1,628 @@
+				CurrentFriend->SetUserLocalAttribute(USER_ATTR_REALNAME, OnlineUser->GetRealName());
+
+				FString PreferredDisplayName;
+				OnlineUser->GetUserAttribute(USER_ATTR_PREFERRED_DISPLAYNAME, PreferredDisplayName);
+				CurrentFriend->SetUserLocalAttribute(USER_ATTR_PREFERRED_DISPLAYNAME, PreferredDisplayName);
+				FString AliasName;
+				OnlineUser->GetUserAttribute(USER_ATTR_ALIAS, AliasName);
+				CurrentFriend->SetUserLocalAttribute(USER_ATTR_ALIAS, AliasName);
+
+				FString DisplayName;
+				CurrentFriend->GetUserAttribute(USER_ATTR_DISPLAYNAME, DisplayName);
+				GEngine->AddOnScreenDebugMessage(-1, 5.f, FColor::Red, "Name is: " + CurrentFriend->GetDisplayName()
+					+ ", with nickname of: " + AliasName);
+			}
+			else
+			{
+				UE_LOG_ONLINE_FRIEND(Error, TEXT("Friendslists not active yet for local user: %d"), InLocalUserNum);
+			}
+		}
+	}
+
+	//It is ok if we don't execute this with presence status, that can be updated through its own callback
+	if (ReadFriendsListDelegate.IsBound() && bWasSuccessful) {
+		ReadFriendsListDelegate.ExecuteIfBound(InLocalUserNum, true, "Default", "");
+	}
+	else
+	{
+		ReadFriendsListDelegate.ExecuteIfBound(InLocalUserNum, false, "Default", FString::Printf(TEXT("Could perform a query, but could not isolate out delegates.")));
+	}
+
+
+	// Clear delegates for the various async calls
+	this->Subsystem->GetUserInterface()->ClearOnQueryUserInfoCompleteDelegate_Handle(0, OnQueryUserInfoCompleteDelegateHandle);
+}
+
+TSharedPtr<FOnlineFriend> FOnlineFriendsInterfaceEpic::GetFriend(int32 LocalUserNum, const FUniqueNetId& FriendId, const FString& ListName)
+{
+	TSharedPtr<FOnlineFriend> Result;
+	if (LocalUserNum < MAX_LOCAL_PLAYERS && Subsystem != NULL)
+	{
+		FEpicFriendsList* FriendsList = FriendsLists.Find(LocalUserNum);
+		if (FriendsList != NULL)
+		{
+			for (int32 FriendIdx = 0; FriendIdx < FriendsList->Friends.Num(); FriendIdx++)
+			{
+
+				if (*FriendsList->Friends[FriendIdx]->GetUserId() == FriendId)
+				{
+					Result = FriendsList->Friends[FriendIdx];
+					break;
+				}
+			}
+		}
+	}
+	return Result;
+}
+
+bool FOnlineFriendsInterfaceEpic::IsFriend(int32 InLocalUserNum, const FUniqueNetId& FriendId, const FString& ListName)
+{
+	if (InLocalUserNum < MAX_LOCAL_PLAYERS && FriendId.IsValid()) {
+		const FOnlineFriendEpic EpicFriend(FUniqueNetIdEpic(FriendId).ToEpicAccountId());
+		return (EpicFriend.GetFriendStatus() != EFriendStatus::NotFriends);
+	}
+	return false;
+}
+
+// ---------------------------------------------
+// EOS SDK Callback functions
+// ---------------------------------------------
+
+void FOnlineFriendsInterfaceEpic::OnEASQueryFriendsComplete(EOS_Friends_QueryFriendsCallbackInfo const* Data)
+{
+	// Make sure the received data is valid on a low level
+	check(Data);
+
+	// To raise the friends query complete delegate the interface itself has to be retrieved from the returned data
+	FLocalUserData* ThisPtr = (FLocalUserData*)(Data->ClientData);
```

Comment:
It's better to use `static_cast<FLocalUserData>(Data->ClientData);` instead of a C-style cast.